contracts/AINFT721.sol

```diff
@@ -0,0 +1,269 @@
+     */
+    function _requireMinted(uint256 tokenId) internal view virtual override {
+        super._requireMinted(tokenId);
+    }
+
+    function tokenURI(
+        uint256 tokenId
+    ) public view override(ERC721) returns (string memory) {
+        _requireMinted(tokenId);
+        return getRecentTokenURI(tokenId);
+    }
+
+    function _getMetadataStorageKey(
+        uint256 tokenId,
+        uint256 version
+    ) internal pure returns (bytes32) {
+        return
+            keccak256(
+                bytes(
+                    string(
+                        abi.encodePacked(
+                            tokenId.toString(),
+                            "AINFT delimeter",
+                            version.toString()
+                        )
+                    )
+                )
+            );
+    }
+
+    function getMetadataStorage(
+        uint256 tokenId
+    ) public view returns (MetadataContainer memory) {
+        //TODO
+        uint256 currentVersion = _tokenURICurrentVersion[tokenId];
+        bytes32 key = _getMetadataStorageKey(tokenId, currentVersion);
+
+        return _metadataStorage[key];
+    }
+
+    function getTokenURICurrentVersion(
+        uint256 tokenId
+    ) public view returns (uint256) {
+        return _tokenURICurrentVersion[tokenId];
+    }
+
+    function getOriginTokenURI(
+        uint256 tokenId
+    ) public view override returns (string memory) {
+        return string(abi.encodePacked(_baseURI(), "/", tokenId.toString()));
+    }
+
+    function getRecentTokenURI(
+        uint256 tokenId
+    ) public view override returns (string memory) {
+        uint256 currentVersion = _tokenURICurrentVersion[tokenId];
+        if (currentVersion == 0) {
+            return getOriginTokenURI(tokenId);
+        } else {
+            bytes32 metadataKey = _getMetadataStorageKey(
+                tokenId,
+                currentVersion
+            );
+            return _metadataStorage[metadataKey].metadataURI;
+        }
+    }
+
+    function getCertainTokenURI(
+        uint256 tokenId,
+        uint256 uriVersion
+    ) public view override returns (string memory) {
+        if (uriVersion == 0) {
+            return getOriginTokenURI(tokenId);
+        } else {
+            bytes32 metadataKey = _getMetadataStorageKey(tokenId, uriVersion);
+            return _metadataStorage[metadataKey].metadataURI;
+        }
+    }
+
+    ////
+    ////
+
+    ////
+    // UPDATE URI(METADATA) FUNCTIONS
+    ////
+
+    function setBaseURI(
+        string memory newBaseURI
+    ) public onlyRole(DEFAULT_ADMIN_ROLE) returns (bool) {
+        require(
+            bytes(newBaseURI).length > 0,
+            "AINFT721::setBaseURI() - Empty newBaseURI"
+        );
+        require(
+            keccak256(bytes(newBaseURI)) != keccak256(bytes(baseURI)),
+            "AINFT721::setBaseURI() - Same newBaseURI as baseURI"
+        );
+
+        baseURI = newBaseURI;
+        return true;
```

Comment:
should emit `BatchMetadataUpdate(start, end)` with counter's help?

Reply:
`BatchMetadataUpdate` is for the NFT marketplaces or other 3rd parties, so I think this should be emitted on here.
If not, we don't have to implement the ERC-4906 at first.